occasional failure from multiprocessing code when running tests

I have seen occasional errors from `sherpa/ui/tests/test_ui_plot.py::test_get_pvalue_plot`. There's an error in multiprocessing, with something  low-level falling over, and you end up with a lot of low-level information and finally

```
Tcl_AsyncDelete: async handler deleted by the wrong thread
```

For more details than I care to know

```
Fatal Python error: Aborted

Current thread 0x00007fd496b14640 (most recent call first):
  Garbage-collecting
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/threading.py", line 247 in __init__
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/queue.py", line 50 in __init__
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/multiprocessing/managers.py", line 387 in create
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/multiprocessing/managers.py", line 209 in _handle_request
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/multiprocessing/managers.py", line 231 in handle_request
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/threading.py", line 946 in run
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/threading.py", line 1009 in _bootstrap_inner
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/threading.py", line 966 in _bootstrap

Thread 0x00007fd494313640 (most recent call first):
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/socket.py", line 293 in accept
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/multiprocessing/connection.py", line 614 in accept
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/multiprocessing/connection.py", line 468 in accept
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/multiprocessing/managers.py", line 189 in accepter
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/threading.py", line 946 in run
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/threading.py", line 1009 in _bootstrap_inner
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/threading.py", line 966 in _bootstrap

Thread 0x00007fd4be829300 (most recent call first):
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/threading.py", line 324 in wait
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/threading.py", line 600 in wait
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/multiprocessing/managers.py", line 176 in serve_forever
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/multiprocessing/managers.py", line 599 in _run_server
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/multiprocessing/process.py", line 108 in run
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/multiprocessing/process.py", line 315 in _bootstrap
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/multiprocessing/popen_fork.py", line 71 in _launch
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/multiprocessing/popen_fork.py", line 19 in __init__
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/multiprocessing/context.py", line 277 in _Popen
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/multiprocessing/process.py", line 121 in start
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/multiprocessing/managers.py", line 562 in start
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/multiprocessing/context.py", line 57 in Manager
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/sherpa/utils/__init__.py", line 3069 in parallel_map
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/sherpa/sim/simulate.py", line 284 in run
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/sherpa/ui/utils.py", line 8655 in get_pvalue_plot
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/sherpa/astro/ui/utils.py", line 10596 in get_pvalue_plot
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/sherpa/ui/tests/test_ui_plot.py", line 2181 in test_get_pvalue_plot
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/_pytest/python.py", line 192 in pytest_pyfunc_call
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/_pytest/python.py", line 1761 in runtest
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/_pytest/runner.py", line 166 in pytest_runtest_call
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/_pytest/runner.py", line 259 in <lambda>
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/_pytest/runner.py", line 338 in from_call
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/_pytest/runner.py", line 258 in call_runtest_hook
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/_pytest/runner.py", line 219 in call_and_report
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/_pytest/runner.py", line 130 in runtestprotocol
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/_pytest/runner.py", line 111 in pytest_runtest_protocol
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/_pytest/main.py", line 347 in pytest_runtestloop
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/_pytest/main.py", line 322 in _main
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/_pytest/main.py", line 268 in wrap_session
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/_pytest/main.py", line 315 in pytest_cmdline_main
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/_pytest/config/__init__.py", line 164 in main
  File "/home/dburke/miniconda3/envs/standalone/lib/python3.10/site-packages/sherpa/__init__.py", line 882 in clitest
  File "/home/dburke/miniconda3/envs/standalone/bin/sherpa_test", line 8 in <module>

Extension modules: numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, sherpa.utils._utils, sherpa.utils._psf, erfa.ufunc, astropy.io.fits._utils, astropy.io.fits.compression, sherpa.estmethods._est_funcs, sherpa.optmethods._saoopt, sherpa.astro.utils._utils, sherpa.astro.utils._pileup, sherpa.models._modelfcts, sherpa.stats._statfcts, matplotlib._c_internal_utils, PIL._imaging, matplotlib._path, kiwisolver._cext, matplotlib._image, group, sherpa.astro.utils._wcs, sherpa.astro.models._modelfcts, stk, numpy.linalg.lapack_lite, sherpa.optmethods._tstoptfct, astropy.table._column_mixins, astropy.table._np_utils, astropy.io.ascii.cparser, yaml._yaml, astropy.utils.xml._iterparser, astropy.time._parse_times, astropy.io.votable.tablewriter, matplotlib.backends._tkagg, PIL._imagingtk (total: 44)
```

and

```
...
../miniconda3/envs/standalone/lib/python3.10/multiprocessing/connection.py:419: in _recv_bytes
    buf = self._recv(4)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <multiprocessing.connection.Connection object at 0x7fd491c37eb0>, size = 4, read = <built-in function read>

    def _recv(self, size, read=_read):
        buf = io.BytesIO()
        handle = self._handle
        remaining = size
        while remaining > 0:
            chunk = read(handle, remaining)
            n = len(chunk)
            if n == 0:
                if remaining == size:
>                   raise EOFError
E                   EOFError

../miniconda3/envs/standalone/lib/python3.10/multiprocessing/connection.py:388: EOFError
```

I've seen this randomly - for instance a python 3.9 build with `pytest` (`pytest-xdist` is installed but not directly in use) and a different test failed, but still with some sort of a crash log. The test cleanup reported something about the async handler

```
___________________________________________________________________________________________________________ test_intproj[sherpa.plot.dummy_backend] ___________________________________________________________________________________________________________
[gw7] linux -- Python 3.9.12 /home/dburke/miniconda3/envs/sherpa-master/bin/python

self = <AutoProxy[Queue] object, typeid 'Queue' at 0x7feba3e71e80>, methodname = 'empty', args = (), kwds = {}

    def _callmethod(self, methodname, args=(), kwds={}):
        '''
        Try to call a method of the referent and return a copy of the result
        '''
        try:
>           conn = self._tls.connection
E           AttributeError: 'ForkAwareLocal' object has no attribute 'connection'

../../miniconda3/envs/sherpa-master/lib/python3.9/multiprocessing/managers.py:802: AttributeError

During handling of the above exception, another exception occurred:

old_numpy_printing = None, override_plot_backend = None

    def test_intproj(old_numpy_printing, override_plot_backend):
        p = plot.IntervalProjection()
        r = p._repr_html_()

        check_empty(r, 'IntervalProjection', nsummary=8)

        x = np.arange(5, 8, 0.5)
        y = np.asarray([2, 3, 4, 5, 4, 3])
        dy = y / 2
        d = Data1D('n n', x, y, staterror=dy)

        m = Const1D()

        fit = Fit(d, m, stat=Chi2())
        fr = fit.fit()
        assert fr.succeeded

        p.prepare(min=1, max=6, nloop=10)
>       p.calc(fit, m.c0)

sherpa/plot/tests/test_plot_notebook.py:281:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
sherpa/plot/__init__.py:2764: in calc
    self.y = numpy.asarray(parallel_map(IntervalProjectionWorker(self.log, par, thawed, fit),
sherpa/utils/__init__.py:3090: in parallel_map
    return run_tasks(procs, err_q, out_q, numcores)
sherpa/utils/__init__.py:2962: in run_tasks
    if not err_q.empty():
<string>:2: in empty
    ???
../../miniconda3/envs/sherpa-master/lib/python3.9/multiprocessing/managers.py:806: in _callmethod
    self._connect()
../../miniconda3/envs/sherpa-master/lib/python3.9/multiprocessing/managers.py:793: in _connect
    conn = self._Client(self._token.address, authkey=self._authkey)
../../miniconda3/envs/sherpa-master/lib/python3.9/multiprocessing/connection.py:507: in Client
    c = SocketClient(address)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

address = b'\x00listener-2315848-0'

    def SocketClient(address):
        '''
        Return a connection object connected to the socket given by `address`
        '''
        family = address_type(address)
        with socket.socket( getattr(socket, family) ) as s:
            s.setblocking(True)
>           s.connect(address)
E           ConnectionRefusedError: [Errno 111] Connection refused

../../miniconda3/envs/sherpa-master/lib/python3.9/multiprocessing/connection.py:635: ConnectionRefusedError
-------------------------------------------------------------------------------------------------------------------- Captured stderr call ---------------------------------------------------------------------------------------------------------------------
Tcl_AsyncDelete: async handler deleted by the wrong thread
=================================================================================================================== short test summary info ===================================================================================================================
```
```

--- a/sherpa/conftest.py
+++ b/sherpa/conftest.py
@@ -559,7 +559,7 @@ def old_numpy_printing():
 PLOT_BACKENDS = [dummy_backend]
 if pylab_backend is not None:
 
-    # over-ride the matplotlib backend from TkAgg to Agg to try
+    # override the matplotlib backend from TkAgg to Agg to try
     # and avoid errors from multiprocessing thanks to the error
     # "Tcl_AsyncDelete: async handler deleted by the wrong thread"
     # See issue #1590.

--- a/sherpa/conftest.py
+++ b/sherpa/conftest.py
@@ -558,6 +558,15 @@ def old_numpy_printing():
 
 PLOT_BACKENDS = [dummy_backend]
 if pylab_backend is not None:
+
+    # over-ride the matplotlib backend from TkAgg to Agg to try
+    # and avoid errors from multiprocessing thanks to the error
+    # "Tcl_AsyncDelete: async handler deleted by the wrong thread"
+    # See issue #1590.
+    #
+    import matplotlib
+    matplotlib.use("agg")
+
     PLOT_BACKENDS.append(pylab_backend)
 
 

